components/Typography/src/MDCFontScaler.m

```
@@ -0,0 +1,324 @@
+        UIContentSizeCategorySmall : @14,
+        UIContentSizeCategoryMedium : @15,
+        UIContentSizeCategoryLarge : @16,
+        UIContentSizeCategoryExtraLarge : @18,
+        UIContentSizeCategoryExtraExtraLarge : @20,
+        UIContentSizeCategoryExtraExtraExtraLarge : @22,
+        UIContentSizeCategoryAccessibilityMedium : @26,
+        UIContentSizeCategoryAccessibilityLarge : @30,
+        UIContentSizeCategoryAccessibilityExtraLarge : @34,
+        UIContentSizeCategoryAccessibilityExtraExtraLarge : @38,
+        UIContentSizeCategoryAccessibilityExtraExtraExtraLarge : @42
+      };
+      _scalingCurve = scalingCurve;
+    }
+  }
+
+  return self;
+}
+
+- (UIFont *)scalableFontWithFont:(UIFont *)font {
+  // If it is available, query the preferredContentSizeCategory.
+  UIContentSizeCategory sizeCategory = GetCurrentSizeCategory();
+
+  UIFont *scaledFont;
+  if (font.mdc_scalingCurve) {
+    scaledFont = [font mdc_scaledFontForSizeCategory:sizeCategory];
+    scaledFont.mdc_scalingCurve = _scalingCurve;
+  } else {
+    // The original font does not have a scaling curve, so create a copy some we aren't
+    // attaching associated objects to system fonts
+    scaledFont = [font copy];
+    scaledFont.mdc_scalingCurve = _scalingCurve;
+    scaledFont = [scaledFont mdc_scaledFontForSizeCategory:sizeCategory];
+  }
+
+  return scaledFont;
+}
+
+- (CGFloat)scaledValueForValue:(CGFloat)value {
+  UIContentSizeCategory defaultSizeCategory = UIContentSizeCategoryLarge;
+  // If it is available, query the preferredContentSizeCategory.
+  UIContentSizeCategory currentSizeCategory = GetCurrentSizeCategory();
+
+  NSNumber *defaultFontSizeNumber = _scalingCurve[defaultSizeCategory];
+  NSNumber *currentFontSizeNumber = _scalingCurve[currentSizeCategory];
+
+  // Guard against broken / incomplete scaling curves by returning self if fontSizeNumber is nil.
+  if (currentFontSizeNumber == nil || defaultFontSizeNumber == nil) {
+    return value;
```

Tip: Please raise an exception if only ONE is `nil`. Having neither content size category is expected, but having only ONE is a programmer error. Programmer error should lead to an NSException being raised.

---

While I agree with your assessment, currently MDC-iOS does NOT throw exceptions.

I believe this is a bigger discussion outside the scope of this PR.

---

Filed issue #6914